store/sqlstore/system_store.go

```diff
@@ -86,3 +87,35 @@ func (s SqlSystemStore) PermanentDeleteByName(name string) (*model.System, *mode
 
 	return &system, nil
 }
+
+// InsertIfExists inserts a given system value if it does not already exist. If a value
+// already exists, it returns the old one, else returns the new one.
+func (s SqlSystemStore) InsertIfExists(system *model.System) (*model.System, *model.AppError) {
+	tx, err := s.GetMaster().Begin()
+	if err != nil {
+		return nil, model.NewAppError("SqlSystemStore.InsertIfExists", "store.sql_system.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	defer finalizeTransaction(tx)
+
+	var origSystem model.System
+	if err := tx.SelectOne(&origSystem, `SELECT * FROM Systems
+		WHERE Name = :Name`,
+		map[string]interface{}{"Name": system.Name}); err != nil && err != sql.ErrNoRows {
+		return nil, model.NewAppError("SqlSystemStore.InsertIfExists", "store.sql_system.get_by_name.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if origSystem.Value != "" {
+		// Already a value exists, return that.
+		return &origSystem, nil
+	}
+
+	// Key does not exist, need to insert.
+	if err := tx.Insert(system); err != nil {
```

Comment:
Can multiple transactions of this type run at the same time? 
Is it possible to get here and have all but one of them fail? 
As far as I understand only one should succeed in writing but all of them should return something.

Reply:
That's a great point ! I read a bit about transaction isolation levels. And it seems that the default isolation levels in most DBs is "read-committed". What we need here is "serializable" isolation level. But unfortunately, our gorp fork does not have [BeginTx](https://golang.org/pkg/database/sql/#DB.BeginTx) which allows you to specify the isolation level.

So I'd need to patch our fork, and then re-vendor the library again.

Thanks for catching this btw. Would have been a nasty bug.